src/meta/processors/job/BalanceJobExecutor.cpp

```diff
@@ -665,174 +439,317 @@ void BalanceJobExecutor::calDiff(const HostParts& hostParts,
+                            }
+                            baton.post();
+                          });
+  baton.wait();
+  return ret;
 }
 
-bool DataBalanceJobExecutor::checkZoneLegal(const HostAddr& source, const HostAddr& target) {
-  VLOG(3) << "Check " << source << " : " << target;
-  auto sourceIter = std::find_if(zoneParts_.begin(), zoneParts_.end(), [&source](const auto& pair) {
-    return source == pair.first;
-  });
+Status ZoneBalanceJobExecutor::buildBalancePlan() {
+  for (std::string& zn : lostZones_) {
+    if (!spaceInfo_.zones_.count(zn)) {
+      return Status::Error("space %s does not have zone %s", spaceInfo_.name_.c_str(), zn.c_str());
+    }
+  }
 
-  if (sourceIter == zoneParts_.end()) {
-    LOG(INFO) << "Source " << source << " not found";
-    return false;
+  std::map<std::string, Zone*> activeZones;
+  std::map<std::string, Zone*> lostZones;
+  for (auto& p : spaceInfo_.zones_) {
+    activeZones.emplace(p.first, &p.second);
   }
+  for (std::string& zn : lostZones_) {
+    auto it = activeZones.find(zn);
+    if (it != activeZones.end()) {
+      lostZones.emplace(it->first, it->second);
+      activeZones.erase(it);
+    }
+  }
+  int32_t activeSize = activeZones.size();
+  if (activeSize < spaceInfo_.replica_) {
+    return Status::Error("Not enough alive zone hold replica");
+  }
+  std::vector<BalanceTask> tasks;
 
-  auto targetIter = std::find_if(zoneParts_.begin(), zoneParts_.end(), [&target](const auto& pair) {
-    return target == pair.first;
+  std::vector<Zone*> sortedActiveZones;
+  sortedActiveZones.reserve(activeZones.size());
+  std::map<std::string, std::vector<Host*>> sortedZoneHosts;
+  std::for_each(activeZones.begin(),
+                activeZones.end(),
+                [&sortedActiveZones, &sortedZoneHosts](std::pair<const std::string, Zone*>& p) {
+                  sortedActiveZones.push_back(p.second);
+                  std::vector<Host*>& hosts = sortedZoneHosts[p.first];
+                  for (auto& ph : p.second->hosts_) {
+                    hosts.push_back(&ph.second);
+                  }
+                  std::sort(hosts.begin(), hosts.end(), [](Host*& l, Host*& r) -> bool {
+                    return l->parts_.size() < r->parts_.size();
+                  });
+                  sortedActiveZones.back()->calPartNum();
+                });
+  std::sort(sortedActiveZones.begin(), sortedActiveZones.end(), [](Zone*& l, Zone*& r) -> bool {
+    return l->partNum_ < r->partNum_;
   });
 
-  if (targetIter == zoneParts_.end()) {
-    LOG(INFO) << "Target " << target << " not found";
-    return false;
-  }
+  auto insertPartIntoZone = [&sortedZoneHosts](Zone* zone, PartitionID partId) -> HostAddr {
+    std::vector<Host*>& sortedHosts = sortedZoneHosts[zone->zoneName_];
+    sortedHosts.front()->parts_.emplace(partId);
+    zone->partNum_++;
+    HostAddr ha = sortedHosts.front()->ha_;
+    for (size_t i = 0; i < sortedHosts.size() - 1; i++) {
```

Comment:
Why don't you loop the sorted hosts and find the first host which meets:
`sortedHosts[i]->parts_.size() < sortedHosts[i + 1]` 
to avoid swap?

Reply:
> Why don't you loop the sorted hosts and find the first host which meets: `sortedHosts[i]->parts_.size() < sortedHosts[i + 1]` to avoid swap?

swap pointers is not performance sensitive and it's easy to understand